### ISSUE TITLE
ODS Investigation

### ISSUE THREAD

**Issue by cohiggins** [bcgov/moodle#25]
Collect information about using ODS/CDATA/Web Services. Contact ODS primary (Patrick? Jason?) 

Aug 12: We'll keep this card active while we collect information.

**Comment by warrenchristian1telus** [comment]
I am troubleshooting CDATA with Justin Choy. We have read access working, and seem to have resolved all the CORS request issues to proceed with testing. Currently, there is no available data to test, and I am getting errors when trying to POST new user data. I don't expect it will take too long to resolve, but not much to see/share at this point.

Moodle Web Services have been configured and enabled in prod namespace, along with an ELM user with privileges for creating service tokens, reading and updating user and course data.

 - Added [custom field(s) in Moodle](https://www.colourmylearning.com/2015/05/how-to-lock-user-profile-fields-and-add-custom-fields-in-moodle/) (IDIR)
- Created a [web service client](https://docs.moodle.org/dev/Creating_a_web_service_client) (ELM)
- Site Admin > Server > Web Service > External Services > Functions 
  - **We will need to determine which functions will we need**
  - [Web Services Documentation](https://moodle-950003-prod.apps.silver.devops.gov.bc.ca/admin/webservice/documentation.php) (on prod - requires Moodle Admin user)
- Created user for ELM Web Service Access (username: elm)
  - Assigned elm to manager role (can create tokens)
- Test adding user data (IDIR field)
  - Only added to admin user for now
- Try core_user_get_users (get all confirmed users - should be everybody)
  - Successful test: All users returned

**Comment by warrenchristian1telus** [comment]
CDATA Query and Create operations are now working.

The API is currently connected to a temporary database, so data and structure are arbitrary. Waiting on ODS connection.

API currently has Update and Delete operations disabled. We may not need these, but we should verify. My concern is that with no Delete ability, the data will grow infinitely. On the other hand, there may be value in having a full record of transactions, and storage is cheap. However, this could (and maybe should) be limited to recording in Moodle, ELM and BI to reduce the possibility of performance issues later, when we have to sift through millions of records to see what's been updated.

**Comment by warrenchristian1telus** [comment]
CDATA Query and Create operations are now working.

The API is currently connected to a temporary database, so data and structure are arbitrary. Waiting on ODS connection.

API currently has Update and Delete operations disabled. We may not need these, but we should verify. My concern is that with no Delete ability the data will grow infinitely. On the other hand, there may be value in having a full record of transactions, and storage is pretty cheap. 

I think we will need to determine data retention policies for ODS / CDATA. Rather than deleting, the system could automatically purge records older than X months, etc.

As the data structure for ODS and data export from PeopleSoft / ELM are still in progress, I've started working on the Moodle export functionality. I have installed [Event Trigger](https://moodle.org/plugins/tool_trigger) in the [prod namespace](https://moodle-950003-prod.apps.silver.devops.gov.bc.ca) which allows site administrators to trigger external actions such as an http request when certain events happen in Moodle.
